app-ios/tutanota/Sources/Offline/IosSqlCipherFade.swift

```diff
@@ -61,4 +72,56 @@ class IosSqlCipherFacade: SqlCipherFacade {
       }
     }
   }
+
+  func vaccumDb() async throws {
+    self.db?.vacuum()
+  }
+
+  /**
+   * We want to lock the access to the "ranges" db when updating / reading the
+   * offline available mail list ranges for each mail list (referenced using the listId).
+   * @param listId the mail list that we want to lock
+   */
+  func lockRangesDbAccess(_ listId: String) async throws {
+    let listIdLock = await concurrentListIdLocks.get(listId)
+    if let listIdLock = listIdLock {
+      await withCheckedContinuation { (continuation: CheckedContinuation<Void, Never>) in
+        let cancellable = listIdLock
+          .first(where: { $0 == .listIdUnlocked })
+          .sink { v in
+            continuation.resume()
+          }
+        self.cancellables.append(cancellable)
```

Comment:
I just had a thought that arrays might be non thread-safe but I think they are CoW anyway

Reply:
They are not, I made them thread safe.